File: 2023/11/ts/main.ts
import { run } from 'src/run.ts';

/** If part 2 test input has completely different input, set this to `true`. */
export const HAS_ALTERNATE = false;

function isBetween(val: number, n1: number, n2: number) {
   return (n1 > val && val > n2) || (n2 > val && val > n1);
}

export function part1(input: string, _isTest: boolean): string {
   const grid: string[] = input.split('\n');
   const galaxies = [];
   const xRowAdd = [];
   const yRowAdd = [];
   let sz = 0;
   for (let x = 0; x < grid[0].length; x++) {
      let flag = false;
      for (let y = 0; y < grid.length; y++) {
         if (grid[y][x] === '#') galaxies[sz++] = [x, y];
         flag ||= grid[y][x] === '#';
      }
      if (!flag) xRowAdd.push(x);
   }
   xdd: for (let y = 0; y < grid.length; y++) {
      for (let x = 0; x < grid[0].length; x++) if (grid[y][x] === '#') continue xdd;
      yRowAdd.push(y);
   }

   let res = 0;
   for (let m = 0; m < sz; m++) {
      for (let n = m + 1; n < sz; n++) {
         const loc1 = galaxies[m];
         const loc2 = galaxies[n];
         const add = xRowAdd.filter((x) => isBetween(x, loc1[0], loc2[0])).length +
            yRowAdd.filter((y) => isBetween(y, loc1[1], loc2[1])).length;
         res += Math.abs(loc1[0] - loc2[0]) + Math.abs(loc1[1] - loc2[1]) + add;
      }
   }

   return res.toString();
}

export function part2(input: string, _isTest: boolean): string {
   const grid: string[] = input.split('\n');
   const galaxies = [];
   const xRowAdd = [];
   const yRowAdd = [];
   let sz = 0;
   for (let x = 0; x < grid[0].length; x++) {
      let flag = false;
      for (let y = 0; y < grid.length; y++) {
         if (grid[y][x] === '#') galaxies[sz++] = [x, y];
         flag ||= grid[y][x] === '#';
      }
      if (!flag) xRowAdd.push(x);
   }
   xdd: for (let y = 0; y < grid.length; y++) {
      for (let x = 0; x < grid[0].length; x++) if (grid[y][x] === '#') continue xdd;
      yRowAdd.push(y);
   }

   let res = 0;
   for (let m = 0; m < sz; m++) {
      for (let n = m + 1; n < sz; n++) {
         const loc1 = galaxies[m];
         const loc2 = galaxies[n];
         const add = xRowAdd.filter((x) => isBetween(x, loc1[0], loc2[0])).length +
            yRowAdd.filter((y) => isBetween(y, loc1[1], loc2[1])).length;
         res += Math.abs(loc1[0] - loc2[0]) + Math.abs(loc1[1] - loc2[1]) + add * 1_000_000 - add;
      }
   }

   return res.toString();
}

if (import.meta.main) {
   run(import.meta.url, part1, part2, HAS_ALTERNATE);
}

File: 2023/11/c/main.c
#include <ctype.h>
#include <helper.h>
#include <main.h>
#include <math.h>
#include <run.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

const int HAS_ALTERNATE = 0;

void parseInput(const char *restrict input, long long ***locations,
                long long **xRow, long long **yRow, int *l_sz, int *x_sz,
                int *y_sz) {
   int i, j, x, y, max_x, max_y;
   int flag;
   char **parsed = strsplit(input, "\n", &max_y);
   max_x = strlen(parsed[0]);

   *l_sz = 0;
   for (x = 0; x < max_x; x++)
      for (y = 0; y < max_y; y++)
         if (parsed[y][x] == '#')
            (*l_sz)++;

   *locations = malloc(*l_sz * sizeof(long long **));
   *xRow = malloc(*l_sz * sizeof(long long *));
   *yRow = malloc(*l_sz * sizeof(long long *));

   i = j = 0;
   for (x = 0; x < max_x; x++) {
      flag = false;
      for (y = 0; y < max_y; y++) {
         if (parsed[y][x] == '#') {
            flag = true;
            (*locations)[j] = malloc(2 * sizeof(long long));
            (*locations)[j][0] = x;
            (*locations)[j][1] = y;
            j++;
         }
      }
      if (!flag) {
         (*xRow)[i++] = x;
      }
   }
   *x_sz = i;

   i = 0;
   for (y = 0; y < max_y; y++) {
      flag = false;
      for (x = 0; x < max_x; x++) {
         if (parsed[y][x] == '#') {
            flag = true;
            break;
         }
      }
      if (!flag) {
         (*yRow)[i++] = y;
      }
      free(parsed[y]);
   }
   free(parsed);
   *y_sz = i;
}

long long lenBetweenLocations(const long long *restrict row, const int sz,
                              const long long x1, const long long x2) {
   long long count = 0;
   for (int i = 0; i < sz; i++) {
      long long x = row[i];
      if ((x1 > x && x > x2) || (x2 > x && x > x1))
         count++;
   }
   return count;
}

char *part1(const char *restrict input) {
   int l_sz, x_sz, y_sz, m, n;
   long long **locations, *xRow, *yRow, *loc1, *loc2, x1, x2, y1, y2, add,
       res = 0;
   parseInput(input, &locations, &xRow, &yRow, &l_sz, &x_sz, &y_sz);

   for (m = 0; m < l_sz; m++) {
      loc1 = locations[m];
      x1 = loc1[0];
      y1 = loc1[1];
      for (n = m + 1; n < l_sz; n++) {
         loc2 = locations[n];
         x2 = loc2[0];
         y2 = loc2[1];
         add = lenBetweenLocations(xRow, x_sz, x1, x2) +
               lenBetweenLocations(yRow, y_sz, y1, y2);
         res += llabs(x1 - x2) + llabs(y1 - y2) + add;
      }
      free(locations[m]);
   }
   free(locations);

   free(xRow);
   free(yRow);
   return numtostr(res);
}

char *part2(const char *restrict input) {
   int l_sz, x_sz, y_sz, m, n;
   long long **locations, *xRow, *yRow, *loc1, *loc2, x1, x2, y1, y2, add,
       res = 0;
   parseInput(input, &locations, &xRow, &yRow, &l_sz, &x_sz, &y_sz);

   for (m = 0; m < l_sz; m++) {
      loc1 = locations[m];
      x1 = loc1[0];
      y1 = loc1[1];
      for (n = m + 1; n < l_sz; n++) {
         loc2 = locations[n];
         x2 = loc2[0];
         y2 = loc2[1];
         add = lenBetweenLocations(xRow, x_sz, x1, x2) +
               lenBetweenLocations(yRow, y_sz, y1, y2);
         res += llabs(x1 - x2) + llabs(y1 - y2) + add * 1000000 - add;
      }
      free(locations[m]);
   }
   free(locations);

   free(xRow);
   free(yRow);
   return numtostr(res);
}

int main(int argc, char *argv[]) {
   return run(argc, argv, part1, part2, HAS_ALTERNATE);
}
